In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [ ]:
#  Import the “products.csv” data set from your “Original Data” folder as df_prods

In [2]:
df_prods = pd.read_csv(r'C:\Users\49157\Downloads\Instacart Basket Analysis\02 Data\Original Data\4.3_orders_products\products.csv', index_col = False)

In [ ]:
# Import“orders_wrangled.csv” data set from your “Prepared Data” folder as df_ords.

In [5]:
df_ords = pd.read_csv(r'C:\Users\49157\Downloads\Instacart Basket Analysis\02 Data\Prepared Data\orders_wrangled.csv', index_col = False)

In [29]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [ ]:
# descriptive statistics for the numeric values in my dataframe.

In [7]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [8]:
# Create a data frame
df_test = pd.DataFrame()

In [9]:
# Create a mixed type columin
df_test['mix'] = ['a', 'b', 1, True]

In [10]:
df_test.head()

mix
0     a
1     b
2     1
3  True

# Conducting consistency checks for products data frame

In [ ]:
# Checking whether a data frame contains any mixed-type columns is as follows:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [ ]:
# Finding missing values in a data frame
df_prods.isnull().sum()

In [11]:
# Creating a subset to view the missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [12]:
# Counting the number of rows and columns
df_prods.shape

(49693, 5)

In [14]:
# Creating a new data frame for df_prods_clean, we want non-missing values in the data frame
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]
df_prods_clean.shape

(49677, 5)

In [15]:
# The following command will look for full duplicates within your dataframe:
df_dups = df_prods_clean[df_prods_clean.duplicated()]
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [17]:
# create a new data frame that doesn’t include the duplicates you just identified using the drop_duplicates() function:

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()
df_prods_clean_no_dups.shape

(49672, 5)

In [24]:
# Creating path
df_prods = pd.read_csv(r'C:\Users\49157\Downloads\Instacart Basket Analysis\02 Data\Original Data\4.3_orders_products\products.csv', index_col = False)


In [26]:
path = r'C:\Users\49157\Downloads\Instacart Basket Analysis'

In [27]:
path

'C:\\Users\\49157\\Downloads\\Instacart Basket Analysis'

In [28]:
# Exporting product- checked data set to csv file
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Conducting consistency checks for orders data frame

In [ ]:
# Step 1 has already been done, please refer to the solutions above

In [30]:
# Step 2 -- Descriptive statistics for the order data frame
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

There is an unusual column in the orders data frame titled Unnamed=0, and I have no idea what this column signifies. The 'Unnamed column' should be investigated further. 

order_number:
Minimum: 1.000000e+00	
Maximum: 1.000000e+02
Observation: The maximum value of 100 for order_number could be unusually high. This might indicate that some users have placed up to 100 orders, which could be possible but should be further investigated for data accuracy. The maximum value  order_number of 100 suggests that some users might have placed up to 100 orders. This could be an outlier or probably an actual high-frequency user behavior. 

In [ ]:
Days_since_prior_order:
The summary statistics in the Days_since_prior_order column suggest there were fewer entries compared to other columns. 
It is important to further investigate why these values are missing and how to handle them appropriately either by imputation using mean or median 
values or remove the missing values completely.


In [ ]:
# Step 3--Check for mixed-type data in your df_ords dataframe.

In [32]:
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
evaluation_set             object
order_number                int64
order_dow                   int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [33]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

C:\Users\49157\AppData\Local\Temp\ipykernel_27772\1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\49157\AppData\Local\Temp\ipykernel_27772\1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\49157\AppData\Local\Temp\ipykernel_27772\1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\49157\AppData\Local\Temp\ipykernel_27772\1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\49157\AppData\Local

# Step 4 -- There are no mixed data types, and we use the function above if there are mixed data types in a data frame.

In [35]:
# Step 5 -- checking for missing values in my df_ords dataframe.
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
evaluation_set                 0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

## Missing values report

## Findings
Checking for missing values in the `df_ords` data frame, the following results were shown:

- `Unnamed: 0`: 0 missing values
- `order_id`: 0 missing values
- `user_id`: 0 missing values
- `order_number`: 0 missing values
- `order_dow`: 0 missing values
- `order_hour_of_day`: 0 missing values
- `days_since_prior_order`: 206,209 missing values

### Explanation for Missing Values in `days_since_prior_order`
The `days_since_prior_order` column has a significant number of missing values (206,209). This column probably represents the number of days since the user's prior order. The missing values can be explained as follows:

1.  When a user places their first order, there is no prior order to reference. Therefore, the value for `days_since_prior_order` would be missing. This could be one reason we have the NaN in this column

2. There could be instances where the information was not recorded properly during data entry, thereby resulting in missing values.

### Proposed Actions to address the missing values
1.  Consider imputing the missing values with statistical measures like the mean or median of the particular column.
2.  Removing and eradicating the missing values completely



In [ ]:
# Step 6--Address the missing values using an appropriate method
First of all, create a subset to see the missing values in the days_since_prior_order column

In [38]:
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]
df_nan

Unnamed: 0  order_id  user_id evaluation_set  order_number  \
0                 0   2539329        1          prior             1   
11               11   2168274        2          prior             1   
26               26   1374495        3          prior             1   
39               39   3343014        4          prior             1   
45               45   2717275        5          prior             1   
...             ...       ...      ...            ...           ...   
3420930     3420930    969311   206205          prior             1   
3420934     3420934   3189322   206206          prior             1   
3421002     3421002   2166133   206207          prior             1   
3421019     3421019   2227043   206208          prior             1   
3421069     3421069   3154581   206209          prior             1   

         order_dow  order_hour_of_day  days_since_prior_order  
0                2                  8                     NaN  
11               2                 11                     NaN  
26               1                 14                     NaN  
39               6                 11                     NaN  
45               3                 12                     NaN  
...            ...                ...                     ...  
3420930          4                 12                     NaN  
3420934          3                 18                     NaN  
3421002          6                 19                     NaN  
3421019          1                 15                     NaN  
3421069          3                 11                     NaN  

[206209 rows x 8 columns]

In [ ]:
# Since I am removing the missing values, let's count the initial rows and columns before the consistency checks

In [39]:
df_ords.shape

(3421083, 8)

In [ ]:
# Creating a new data frame for df_ords_clean, we want non-missing values in the orders data frame

In [41]:
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]
df_ords_clean.shape

(3214874, 8)

## Addressing Missing Values

### Approach used Used
To address the missing values in the `days_since_prior_order` column, I used the method of removing the missing values

### Explanation for Chosen Method
Removing missing values is a practical approach to ensure data integrity, simplify analysis, and improve model performance. However, it is essential to weigh the pros and cons and consider the specific context of the analysis before deciding to remove missing data. On the other hand, if we decide to leave the missing values, we can also make use of the Imputation method, thereby inputting the mean or median of the values in the days_since_prior_order column.

### Verification
After applying this method, I verified that there are no remaining missing values in the `days_since_prior_order` column. This ensures the dataset is complete and ready for further analysis.

In [ ]:
# Step 7--check for duplicate values in your df_ords data.
# The following command will look for full duplicates within my orders data frame:

In [45]:
df_dups = df_ords_clean[df_ords_clean.duplicated()]
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, evaluation_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

In [ ]:
# We do not have duplicates in the order data frame, so to check the count of rows after the missing values were removed, we got 3214874

In [47]:
df_ords_clean.shape

(3214874, 8)

In [ ]:
# We use this function, if there wer duplicates found

In [50]:
df_ords_clean_no_dups = df_ords_clean.drop_duplicates()
df_ords_clean_no_dups.shape

(3214874, 8)

## As no duplicates were found, the value remains the same., but we address duplicates using this approach stated below.

## Addressing Duplicates

### Approach Used
To address the duplicates in the data frame, I used the `drop_duplicates` method provided by pandas. This method effectively removes duplicate rows from the data frame, ensuring that each row is unique.

### Explanation for Chosen Method
The rationale behind this choice is based on the following considerations:

1. Duplicate rows can distort the analysis and lead to inaccurate conclusions. Removing duplicates ensures that each observation in the dataset is unique, maintaining data reliability and integrity.

2. Duplicates can complicate data analysis and interpretation. By removing them, the dataset becomes cleaner and easier to work with, allowing for more straightforward and meaningful analysis.

3. Removing duplicates optimizes the dataset for better performance and efficiency in subsequent analyses.

### Verification
After applying the `drop_duplicates` method, there were no remaining duplicate rows in the data frame. This ensures the dataset is clean and ready for further analysis.

In [ ]:
# Exporting orders data frame, but first of all, we have to create a path

In [54]:
df_ords = pd.read_csv(r'C:\Users\49157\Downloads\Instacart Basket Analysis\02 Data\Original Data\4.3_orders_products\orders.csv', index_col = False)

In [55]:
path = r'C:\Users\49157\Downloads\Instacart Basket Analysis'
path

'C:\\Users\\49157\\Downloads\\Instacart Basket Analysis'

In [ ]:
# Now let us export

In [56]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))